Experience Control: Improve accessibility and minor refactoring

yashwin opened the pull request:
Closes https://github.com/Automattic/automattic-for-agencies-dev/issues/2112

## Proposed Changes

We copied the original implementation of the feedback component on this [PR](https://github.com/Automattic/wp-calypso/pull/102700), which introduced Experience Control to the Automattic package, but it wasn't accessibility-friendly.

In this PR, we are addressing the same issue along with the other feedback we received from @mirka.

## Why are these changes being made?

* To address some accessibility missed along with some refactoring.

## Testing Instructions

* Follow the instructions on the original PR: https://github.com/Automattic/wp-calypso/pull/102700 and everything should work as expected. 
* Verify the accessibility changes work as expected: 

1) VoiceOver when the emojis are focused. You can use System settings > Accessibility > VoiceOver
2) You can use the keyboard to navigate between emojis



https://github.com/user-attachments/assets/5f28c4b1-dcfa-43c5-bb0b-05b260de737c




## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

yashwin committed:
Experience Control: Improve accessibility and minor refactor

madebynoam reviewed (approved):
Great details! Ty, @yashwin ! Lgtm!

mirka commented:
Thanks for the quick follow-up! I'm clarifying a few things first in the Figma thread (dhW7zYBamXJIeJIE3wy5f4-fi-111_846#1193979809) since this is the first time I'm seeing this discussion.

yashwin committed:
update type

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
 	GOOD = 'good',
 	NEUTRAL = 'neutral',
 	BAD = 'bad',
 }
```
I wonder if the overhead of using an enum here is unnecessary. `type ExperienceValue = 'good' | 'neutral' | 'bad'` would be sufficiently safe — we could just pass plain strings to the props and it would be typechecked. (If it matters, we also just use unions for things like this in `@wordpress/components`.)

And a small nit on the type name —  everything is a type so I find "Type" suffixes a bit redundant in types 😂 Maybe `ExperienceValue` is more descriptive here.

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
+	ariaLabel,
+}: ExperienceControlOptionProps ) => (
+	<label
+		className={ clsx( 'experience-control__option', className, {
```
We'll likely convert this to CSS Modules soon (#103004), but in the meantime let's add a namespace prefix:

```suggestion
		className={ clsx( 'a8c-experience-control__option', className, {
```

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
+	value: ExperienceType;
+	icon: JSX.Element;
+	ariaLabel: string;
+};
```
Consolidating the docs to TSDocs (https://github.com/Automattic/wp-calypso/pull/102700#discussion_r2066115316) means to move the prop descriptions that you wrote for the README into the type definitions, so they can be extracted programmatically ([example](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/input-control/types.ts)). You'll probably want to make a separate `types.ts` file if it bloats the main component file.

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
+	value: string;
+	name: string;
+	ariaLabel: string;
+}
```
We want to have an open/closed pattern for the APIs, and we want to maintain consistency with the underlying HTML element API as much as possible. So in this context we don't want a closed set of props, but rather the entire set of `React.ComponentProps< 'input' >`, plus any other additional props we need to add.

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
+			type="radio"
+			className="experience-control__radio"
+			checked={ checked }
+			onChange={ onClick }
```
Why are we passing an `onClick` to the `onChange`? The types won't match if defined properly (right now it's defined as `() => void`, which is misleading).

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
-			'is-selected': isSelected,
+	value,
+	name,
+	ariaLabel,
```
Much of this can be simplified by spreading `...restProps` to the `input` directly.

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
+			name={ name }
+			aria-label={ ariaLabel }
+		/>
+		<div className="experience-control__option-content">{ children }</div>
```
I'm a bit hesitant to use `children` like this, because radio inputs don't have children and it's not obvious from the consumer how a radio component would render a `children` prop. Functionally, it's being rendered as a `label` here. Would it be more straightforward if we just called it the `label`?

```jsx
<label>
  <input type="radio" />
  { label }
</label>
```

And then for example in an NPS-style numerical rating use case, the `label` prop would be passed something like `<StyledNumeral>1</StyledNumeral>`. For use cases where `label` is not a string, it would be something like `<Emoticon aria-label="good" />`.

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
+}: ExperienceControlOptionProps ) => (
+	<label
+		className={ clsx( 'experience-control__option', className, {
+			'is-selected': checked,
```
This extra modifier class is unnecessary since it can be done directly in CSS (e.g. `label:has([checked])`).

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
+			</div>
+		</BaseControl>
+	);
+};
```
Sorry I forgot we need a bit more special handling for radio groups:

```diff
diff --git a/packages/components/src/experience-control/index.tsx b/packages/components/src/experience-control/index.tsx
index 037eef046c4..a4b58d88c9e 100644
--- a/packages/components/src/experience-control/index.tsx
+++ b/packages/components/src/experience-control/index.tsx
@@ -1,4 +1,4 @@
-import { BaseControl, useBaseControlProps } from '@wordpress/components';
+import { BaseControl, useBaseControlProps, VisuallyHidden } from '@wordpress/components';
 import clsx from 'clsx';
 import { useTranslate } from 'i18n-calypso';
 import { IconBad, IconGood, IconNeutral } from './icons';
@@ -54,20 +54,34 @@ const ExperienceControlOption = ( {
 	</label>
 );
 
-interface ExperienceControlBaseProps {
-	children: React.ReactNode;
-	label?: string;
-	help?: string;
-}
+type ExperienceControlBaseProps = Omit<
+	React.ComponentProps< typeof BaseControl >,
+	'__nextHasNoMarginBottom' | 'as'
+>;
 
-const ExperienceControlBase = ( { children, ...props }: ExperienceControlBaseProps ) => {
-	const { baseControlProps } = useBaseControlProps( props );
+const ExperienceControlBase = ( {
+	children,
+	className,
+	hideLabelFromVision,
+	label,
+	...props
+}: ExperienceControlBaseProps ) => {
+	const { baseControlProps, controlProps } = useBaseControlProps( props );
 
 	return (
-		<BaseControl className="experience-control" { ...baseControlProps }>
-			<div className="experience-control__options" role="radiogroup">
+		<BaseControl
+			__nextHasNoMarginBottom
+			className={ clsx( 'experience-control', className ) }
+			{ ...baseControlProps }
+		>
+			<fieldset { ...controlProps } className="experience-control__options">
+				{ hideLabelFromVision ? (
+					<VisuallyHidden as="legend">{ label }</VisuallyHidden>
+				) : (
+					<BaseControl.VisualLabel as="legend">{ label }</BaseControl.VisualLabel>
+				) }
 				{ children }
-			</div>
+			</fieldset>
 		</BaseControl>
 	);
 };
```
- This should now associate everything correctly, including the radio group name and the `help` text. (The current implementation associates neither, which can be confirmed in the accessibility tree.)
- Enables the `hideLabelFromVision` functionality, which is a standard API on our controls.
- Forces the `__nextHasNoMarginBottom` prop, which is recommended for new components, and is currently logging a console warning.

We'll also need to add margin/padding/border reset styles for the `fieldset`.

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
 		},
 	];
 
+	const radioGroupName = `experience-control-${ label.toLowerCase().replace( /\s+/g, '-' ) }`;
```
This isn't safe against collisions, and also `name` needs to be overridable by the consumer. So let's take a `name` prop, and default it to a unique id (`React.useId` or `useInstanceId` from WP).

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
-			value: Experience.GOOD,
+			value: ExperienceType.GOOD,
 			icon: <IconGood />,
+			ariaLabel: translate( 'Rate as good experience' ),
```
Can we use `@wordpress/react-i18n` for translation? (Context: p1746016454012249-slack-CK1QYS6P6)

mirka replied:
The label "Rate as ___ experience" is a bit redundant, because the radio group label is already going to provide the context ("How was your experience?", "Rate your experience", etc).

I think the strings can just be "Good"/"Neutral"/"Bad".

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
-					isSelected={ selectedExperience === option.value }
-					onClick={ () => handleChange( option.value ) }
+					checked={ value === option.value }
+					onClick={ () => onChange( option.value ) }
```
We should generally prefer `onChange` over `onClick` for state changes because `onClick` can fire even when the state hasn't changed (like clicking the same option multiple times).

So that would usually look something like:

```suggestion
					onChange={ ( event: React.ChangeEvent< HTMLInputElement > ) =>
						onChange( event.target.value as ExperienceType )
					}
```

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
 					className={ `is-${ option.value }` }
-					isSelected={ selectedExperience === option.value }
-					onClick={ () => handleChange( option.value ) }
+					checked={ value === option.value }
```
Currently this code is written like it will only be used in controlled mode, but the stories are in uncontrolled mode so it doesn't match. Should we keep it in controlled mode only for now, and change the stories? Otherwise the stories just look broken.

mirka commented on packages/components/src/experience-control/style.scss:
```diff
@@ -27,60 +27,65 @@ $experience-types: (
-	display: flex;
-	flex-direction: column;
-	gap: 12px;
+	.components-base-control__label {
+		@include heading-medium;
+		margin-block-end: 12px;
+		text-transform: uppercase;
+		color: var(--color-text);
+	}
```
We don't want to override the Core label styles, these are standardized across components. Is it possible to update the A4A screen so it uses standard Core components/styling?

mirka commented on packages/components/src/index.ts:
```diff
@@ -66,6 +66,6 @@ export { WooCommerceWooLogo } from './logos/woocommerce-woo-logo';
 export { default as DotPager } from './dot-pager';
 export { default as EmbedContainer } from './embed-container';
-export { ExperienceControl } from './experience-control';
+export { ExperienceControl, ExperienceType } from './experience-control';
```
We generally don't want to expose types directly if they can be extracted from the component types (e.g. `React.ComponentProps< typeof ExperienceControl >[ 'value' ]`).

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
 	);
 }
+
 ExperienceControl.Base = ExperienceControlBase;
 ExperienceControl.Option = ExperienceControlOption;
```
From what I gathered from the Figma thread, we don't have a good idea of what abstractions will be appropriate yet. We'll want to work closely with the people needing the initial use cases, so they don't start customizing it locally. Let's keep the subcomponents internal for now?

```suggestion
```

mirka commented on packages/components/src/experience-control/index.stories.tsx:
```diff
@@ -12,33 +13,39 @@ type Story = StoryObj< typeof ExperienceControl >;
 };
 
 // Example of using the Base component directly
 export const CustomBase: Story = {
 	render: () => (
 		<ExperienceControl.Base label="Custom Experience Control">
-			<ExperienceControl.Option isSelected onClick={ () => {} }>
+			<ExperienceControl.Option
+				checked
+				onClick={ fn() }
+				value="1"
+				name="experience-control"
+				ariaLabel="Rate as good experience"
+			>
 				1
 			</ExperienceControl.Option>
 		</ExperienceControl.Base>
```
Let's remove this story for now, and we can come back to custom examples when we're ready?

mirka commented on packages/components/src/experience-control/index.stories.tsx:
```diff
@@ -1,8 +1,9 @@
 const meta: Meta< typeof ExperienceControl > = {
-	title: 'Components/ExperienceControl',
+	title: 'packages/components/ExperienceControl',
 	component: ExperienceControl,
```
```suggestion
	component: ExperienceControl,
	args: { onChange: fn() },
```

FYI we can set all the `onChange` spies in the `meta`.

mirka commented on packages/components/src/experience-control/icons.tsx:
```diff
@@ -1,47 +1,32 @@
+			fillRule="evenodd"
+			clipRule="evenodd"
+			d="M5.99452 13.0808C6.89004 13.0808 7.616 12.3549 7.616 11.4594C7.616 10.5638 6.89004 9.83789 5.99452 9.83789C5.09901 9.83789 4.37305 10.5638 4.37305 11.4594C4.37305 12.3549 5.09901 13.0808 5.99452 13.0808ZM18.0054 13.0808C18.901 13.0808 19.6269 12.3549 19.6269 11.4594C19.6269 10.5638 18.901 9.83789 18.0054 9.83789C17.1099 9.83789 16.384 10.5638 16.384 11.4594C16.384 12.3549 17.1099 13.0808 18 ...
+			fill="#255C33"
```
Since this scatters the color variable management, could we maybe use `currentColor` for the fills here, and control the actual color with the `color` property in the stylesheet?

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,97 +1,126 @@
+}
+
+const ExperienceControlBase = ( { children, ...props }: ExperienceControlBaseProps ) => {
+	const { baseControlProps } = useBaseControlProps( props );
+
+	return (
+		<BaseControl className="experience-control" { ...baseControlProps }>
+			<div className="experience-control__options" role="radiogroup">
+				{ children }
+			</div>
+		</BaseControl>
+	);
+};
 
-export function ExperienceControl( {
-	label,
-	onChange,
-	selectedExperience,
-	helpText,
-}: {
+interface ExperienceControlProps {
 	label: string;
-	onChange: ( experience: string ) => void;
-	selectedExperience: string;
-	helpText?: string;
-} ) {
-	const handleChange = ( experience: string ) => {
-		onChange( experience );
-	};
+	onChange: ( experience: ExperienceType ) => void;
+	value: ExperienceType;
+	help?: string;
+}
```
Some of these can be `Pick<>`ed off of the types we already defined, which will keep things DRY (especially when they include TSDoc descriptions).

ciampo commented:
Given that we're looking to release a new version of the components package soon (https://github.com/Automattic/wp-calypso/pull/103160), could it be wiser to remove `ExperienceControl` from the package exports so that we don't release a WIP component? cc @mirka @simison

mirka commented:
> Given that we're looking to release a new version of the components package soon (#103160), could it be wiser to remove `ExperienceControl` from the package exports so that we don't release a WIP component?

As a quick fix (?) I [moved](https://github.com/Automattic/wp-calypso/pull/103146/commits/ae5158ee505096d9f8745ad0bbf5843fd2967d27) it to the `Unaudited` section in Storybook.

yashwin committed:
address comments

yashwin commented:
@mirka:

Thanks for the feedback! 🙇 

I have addressed all the comments:https://github.com/Automattic/wp-calypso/pull/103000/commits/b49d4a8c13a3963de3fe43ac5bb8c74f6047d504. Could you please take another look?

aduth commented on packages/components/src/experience-control/style.scss:
```diff
@@ -26,61 +26,68 @@ $experience-types: (
+	gap: 8px;
+	margin: 0;
+    padding: 0;
+    border: none;
```
There's mixed indentation here. This is something that CI should be catching, but it's missing from our checks. This is being fixed in #103396, so you'll want to fix and/or rebase your branch after that's merged to avoid introducing build failures on `trunk`.

```suggestion
	gap: 8px;
	margin: 0;
	padding: 0;
	border: none;
```

yashwin replied:
Thanks for letting me know. This is mostly done by Cursor. Sorry for missing this earlier. Fixed it here: https://github.com/Automattic/wp-calypso/pull/103000/commits/3229a5b6496985e6316b404474d6c2899bf4db16

yashwin committed:
Fix indentation

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,98 +1,132 @@
+	<label className={ clsx( 'a8c-experience-control__option', className ) }>
+		<input
+			type="radio"
+			className="a8c-experience-control__radio"
```
CSS is being written to hide this visually, but that can be done with a component like [`VisuallyHidden`](https://wordpress.github.io/gutenberg/?path=/docs/components-visuallyhidden--docs).

mirka commented on packages/components/src/experience-control/index.tsx:
```diff
@@ -1,98 +1,132 @@
+	ariaLabel,
+	...restProps
+}: ExperienceControlOptionProps ) => (
+	<label className={ clsx( 'a8c-experience-control__option', className ) }>
+		<input
+			type="radio"
+			className="a8c-experience-control__radio"
+			aria-label={ ariaLabel }
+			{ ...restProps }
+		/>
+		<div className="a8c-experience-control__option-label">{ label }</div>
+	</label>
```
The use of `label` and `aria-label` here are a bit unclear to me. `label` implies a string, and it's non-obvious whether `label` or `aria-label` is going to take precedence as the actual label of the input. Everything being wrapped in a `<label>` element also strongly implies that `label` is expected to be a text node, or at least have a text representation.

I can think of a few ways around this weirdness, for example we can expect the consumer to pass in an already accessibly-labeled `label` node, instead of a separate aria label that is passed on to the `input`.

yashwin replied:
Agree.

Do you think we can use `icon` instead of `label` something like:

```
const ExperienceControlOption = ( {
	className,
	icon,
	ariaLabel,
	...restProps
}: ExperienceControlOptionProps ) => (
	<label className={ clsx( 'a8c-experience-control__option', className ) }>
		<VisuallyHidden>
			<input type="radio" aria-label={ ariaLabel } { ...restProps } />
		</VisuallyHidden>
		<div className="a8c-experience-control__option-icon" aria-hidden="true">
			{ icon }
		</div>
	</label>
);
```

mirka replied:
Sure, that's probably fine for now, since we only have this one use case. In the future when this abstraction needs to support more use cases, it may be too limiting since it won't necessarily be an icon (e.g. an NPS selector where it's just plain text numbers). But we don't have to overthink it at this moment.

mirka reviewed (approved):
I had some minor thoughts about implementation, but the accessibility issues are fixed 🎉

None of the notes are blockers, but we should probably fix the Storybook controlled mode issue since the stories currently look broken.

mirka commented:
<!-- linear:replyTo:2896102878 -->

Yes, go ahead 👍

yashwin committed:
replace label with icon and use VisuallyHidden to hide the input

mirka committed:
Fix Storybook crash

yashwin committed:
fix the Storybook controlled mode

yashwin commented:
Fixed the Storybook


<img width="1096" alt="Screenshot 2025-06-10 at 12 12 11 PM" src="https://github.com/user-attachments/assets/63cf020c-b147-4f76-b71a-d999cfa31932" />
